File: 백준/Gold/17135. 캐슬 디펜스/캐슬 디펜스.py
import sys
input = sys.stdin.readline
from itertools import combinations

n,m,d = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]



enemies = []
for i in range(n):
    for j in range(m):
        if arr[i][j] == 1:
            enemies.append((i,j))

archer_positions = list(combinations([(n, j) for j in range(m)], 3))
answer = 0

def find_target(x, y, area):
    min_dist = d + 1
    target = (-1, -1)
    for i in range(n):
        for j in range(m):
            if area[i][j] == 1:
                dist = abs(x - i) + abs(y - j)
                if dist <= d:
                    # 거리 우선, 왼쪽 우선으로 갱신
                    if dist < min_dist or (dist == min_dist and j < target[1]):
                        min_dist = dist
                        target = (i, j)
    return target

#적 이동
def move_enemies(area):
    new_area = [[0] * m for _ in range(n)]
    for i in range(n - 1):
        for j in range(m):
            if area[i][j] == 1:
                new_area[i + 1][j] = 1
    return new_area


for l in archer_positions:
    area = [row[:] for row in arr]
    kill = 0
    while True:
        targets = set()

        for x,y in l:
            target = find_target(x,y,area)
            if target != (-1,-1):
                targets.add(target)
        for x,y in targets:
            if area[x][y] == 1:
                area[x][y] = 0
                kill +=1
        area = move_enemies(area)
        if sum(sum(row) for row in area) == 0:
            break
    answer = max(answer, kill)
print(answer)

File: 백준/Gold/17135. 캐슬 디펜스/캐슬 디펜스.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M, D;
    static int[][] map;       // 원본 맵
    static List<int[]> enemies = new ArrayList<>(); // 초기 적 좌표 저장용 (선택)
    static List<int[]> combi = new ArrayList<>();
    static int min = 0;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        // N: 행, M: 열, D: 공격 거리
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken()); // 행 수
        M = Integer.parseInt(st.nextToken()); // 열 수
        D = Integer.parseInt(st.nextToken()); // 공격 거리

        map = new int[N][M];


        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
                
                // 필요 시 적의 초기 위치 저장
                if (map[i][j] == 1) {
                    enemies.add(new int[]{i, j});
                }
            }
        }
        //궁수 위치 조합
        combination(0,0,new int[3]);
        //조합별로 공격 최댓값 구하기
        min = 0;
        for (int[] soldier : combi) {
            attack(soldier);
        }
        System.out.println(min);


    }
    public static void combination(int L, int start, int[] visited){
        if(L==3){
            combi.add(visited.clone());
            return;
        }
        for(int i=start;i<M;i++){
            visited[L]= i;
            combination(L+1, i+1, visited);
        }
    }
    public static void attack(int[] soldier){
        int[][] arr = new int[N][M];
        for (int i = 0; i < N; i++) {
            arr[i] = map[i].clone();
        }
    
        int kill = 0;
        for (int t = 0; t < N; t++) {
            Set<String> target = new HashSet<>();
    
            for (int s : soldier) {
                int minDist = D + 1;
                int row = -1, col = -1;
    
                for (int i = N - 1 - t; i >= 0; i--) {
                    for (int j = 0; j < M; j++) {
                        if (arr[i][j] == 1) {
                            int dist = Math.abs(N - t - i) + Math.abs(s - j);
                            if (dist <= D) {
                                if (dist < minDist || (dist == minDist && j < col)) {
                                    row = i;
                                    col = j;
                                    minDist = dist;
                                }
                            }
                        }
                    }
                }
    
                if (row != -1) {
                    target.add(row + "," + col);
                }
            }
    
            // 적 제거
            for (String pos : target) {
                String[] parts = pos.split(",");
                int r = Integer.parseInt(parts[0]);
                int c = Integer.parseInt(parts[1]);
                arr[r][c] = 0;
            }
    
            kill += target.size();
        }
    
        min = Math.max(min, kill); // 조합별 최대 킬 수 반영
    }
    
}
